flow/internal/workflow.go

```diff
@@ -11,18 +11,58 @@ import (
 	"github.com/PeerDB-io/peerdb/flow/shared"
 )
 
-func GetWorkflowStatus(ctx context.Context, temporalClient client.Client, workflowID string) (protos.FlowStatus, error) {
+func getWorkflowStatusFromTemporal(ctx context.Context, temporalClient client.Client, workflowID string) (protos.FlowStatus, error) {
 	res, err := temporalClient.QueryWorkflow(ctx, workflowID, "", shared.FlowStatusQuery)
 	if err != nil {
 		slog.Error("failed to query status in workflow with ID "+workflowID, slog.Any("error", err))
-		return protos.FlowStatus_STATUS_UNKNOWN,
-			fmt.Errorf("failed to query status in workflow with ID %s: %w", workflowID, err)
+		return protos.FlowStatus_STATUS_UNKNOWN, fmt.Errorf("failed to query status in workflow with ID %s: %w", workflowID, err)
 	}
 	var state protos.FlowStatus
 	if err := res.Get(&state); err != nil {
 		slog.Error("failed to get status in workflow with ID "+workflowID, slog.Any("error", err))
-		return protos.FlowStatus_STATUS_UNKNOWN,
-			fmt.Errorf("failed to get status in workflow with ID %s: %w", workflowID, err)
+		return protos.FlowStatus_STATUS_UNKNOWN, fmt.Errorf("failed to get status in workflow with ID %s: %w", workflowID, err)
 	}
 	return state, nil
 }
+
+func GetWorkflowStatus(ctx context.Context, pool shared.CatalogPool,
+	temporalClient client.Client, workflowID string,
+) (protos.FlowStatus, error) {
+	var flowStatus protos.FlowStatus
+	err := pool.QueryRow(ctx, "SELECT status FROM flows WHERE workflow_id = $1", workflowID).Scan(&flowStatus)
+	if err != nil || flowStatus == protos.FlowStatus_STATUS_UNKNOWN {
+		if err != nil {
+			slog.Error("failed to get status for flow from catalog, will fall back to temporal",
+				slog.Any("error", err),
+				slog.String("flowID", workflowID))
+		} else if flowStatus == protos.FlowStatus_STATUS_UNKNOWN {
+			slog.Info("flow status from catalog is unknown, will fall back to temporal",
+				slog.String("flowID", workflowID))
+		}
+
+		// this should only trigger for existing mirrors once
+		status, tctlErr := getWorkflowStatusFromTemporal(ctx, temporalClient, workflowID)
+		if tctlErr != nil {
+			return status, tctlErr
+		}
+		// only update the catalog if the status is COMPLETED, this a long state, maintenance wfs do not alter these and we can assume
+		// that any changes post read-through cache happen only on new workers
+		if err == nil && status == protos.FlowStatus_STATUS_COMPLETED {
+			return UpdateFlowStatusInCatalog(ctx, pool, workflowID, status)
+		}
```

Comment:
@serprex we can assume here that resyncs for Completed flows only happen on upgraded workers

Reply:
TERMINATING we'll get stuck on since we won't pick up when it gets to TERMINATED, it's okay to query